varats/varats/data/reports/llvm_coverage_report.py

```diff
@@ -0,0 +1,708 @@
+    start: RegionStart
+    end: RegionEnd
+    count: int
+    kind: CodeRegionKind
+    function: str
+    parent: tp.Optional[CodeRegion] = None
+    childs: tp.List[CodeRegion] = field(default_factory=list)
+
+    @classmethod
+    def from_list(cls, region: tp.List[int], function: str) -> CodeRegion:
+        """Instantiates a CodeRegion from a list."""
+        if len(region) > 5:
+            assert region[5:7] == [
+                0, 0
+            ]  # Not quite sure yet what the zeros stand for.
+        return cls(
+            start=RegionStart(line=region[0], column=region[1]),
+            end=RegionEnd(line=region[2], column=region[3]),
+            count=region[4],
+            kind=CodeRegionKind(region[7]),
+            function=function,
+        )
+
+    def iter_breadth_first(self) -> tp.Iterator[CodeRegion]:
+        """Yields childs breadth_first."""
+        todo = deque([self])
+
+        while todo:
+            node = todo.popleft()
+            childs = list(node.childs)
+            todo.extend(childs)
+            yield node
+
+    def iter_preorder(self) -> tp.Iterator[CodeRegion]:
+        """Yields childs in preorder."""
+        yield self
+        for child in self.childs:
+            for x in child.iter_preorder():
+                yield x
+
+    def iter_postorder(self) -> tp.Iterator[CodeRegion]:
+        """Yields childs in postorder."""
+        for child in self.childs:
+            for x in child.iter_postorder():
+                yield x
+        yield self
+
+    def has_parent(self) -> bool:
+        if self.parent is None:
+            return False
+        return True
+
+    def is_covered(self) -> bool:
+        return self.count > 0
+
+    def is_subregion(self, other: CodeRegion) -> bool:
+        """Tests if the 'other' region fits fully into self."""
+        start_ok = False
+        end_ok = False
+
+        if self.start.line < other.start.line:
+            start_ok = True
+        elif self.start.line == other.start.line:
+            start_ok = self.start.column < other.start.column
+
+        if self.end.line > other.end.line:
+            end_ok = True
+        elif self.end.line == other.end.line:
+            end_ok = self.end.column > other.end.column
+
+        return start_ok and end_ok
+
+    def insert(self, region: CodeRegion) -> None:
+        """
+        Inserts the given code region into the tree.
+
+        The new regions must not exist yet and must not overlap
+        """
+        if not self.is_subregion(region):
+            raise ValueError("The given region is not a subregion!")
+        if region in self:
+            raise ValueError("The given region exists already!")
+
+        # Find the right child to append to
+        # Should be the first code region where region is a subregion
+        # when traversing the tree in postorder
+        for node in self.iter_postorder():
+            if node.is_subregion(region):
+                if len(node.childs) > 0:
```

Comment:
```suggestion
                if node.childs:
```
This should be enough, as empty lists are false